### ISSUE TITLE
Memory during training (LoRA)

### ISSUE THREAD

**Issue by simap** [ml-explore/mlx-lm#828]
Heya, first thanks for the awesome library! I'm a newb and this is making it relatively easy.

I was playing around with doing some tuning on a model and kept hitting memory walls (memory pressure to the roof, sometimes hard crashes). What was weird is that the memory kept growing, each val or iteration. Saw some other issues or mentions about KV cache with these kind of symptoms, but that was all for generation, and I don't see anything for that in the code on the training side.

Saw this:
https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/tuner/trainer.py#L214
And I poked around and found a bunch more limit setters. 

Anyway I ended up hacking this in (below), haven't tuned it really, but I think setting these limits to lower values (or exposing them) would really help people out. Without these I was hitting hard walls.

e.g. with some models and lora settings I was going OOM/hard crashing on 48G ram, seeing python at 36G and memory pressure in the red. 

With the same model and lora settings, if I drop wired_limit a little, like to 90%, everything is stable. Still high memory use, but no reboots.

If I drop cache_limit way down or set to zero, my memory never climbs, no crashes or memory pressure, and I don't see any performance impact in my limited testing. 

Setting memory_limit too low or even zero doesn't cause it to fail, but does run slower, at about 80% speed. 

I can even set all 3 to zero and still skip along at 80% speed. Maybe not ideal for a training server farm, but could be useful if you are trying to use the machine at the same time and consume minimal resources while running some background training? Anyway it was surprising and interesting that it didn't outright fail or crawl.

So around [here](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/tuner/trainer.py#L214) at the top of `train()`

```python
    if mx.metal.is_available():
        wired_limit = mx.metal.device_info()["max_recommended_working_set_size"]
        # max_recommended_working_set_size is too large
        wired_limit = int(wired_limit * 0.9)
        print(f"Setting wired limit to {wired_limit}")
        mx.set_wired_limit(wired_limit)

        memory_limit = int(wired_limit * 1.0) #docs say this defaults to 1.5x
        print(f"Setting memory limit to {memory_limit}")
        mx.set_memory_limit(memory_limit)

        cache_limit = int(wired_limit * 0.5) #docs say this defaults to memory_limit
        print(f"Setting cache limit to {cache_limit}")
        mx.set_cache_limit(cache_limit)
```

**Comment by kru2710shna** [comment]
Hi @simap  
Your diagnosis lines up with what I'm seeing in the tuner code path. The MLX buffer cache and wired-limit defaults are tuned for steady-state inference, where the memory profile is bounded by KV cache and predictable. LoRA training adds activations + gradients + optimizer state on top, and the cache policy doesn't bound that growth tightly enough, which is why memory creeps up per iteration and eventually. Your hack at the top of `train()` is essentially the right shape.

I'd like to pick this up. Background: I work on QLoRA / on-device fine-tuning for sub-3B models, so this lives in my area. 
Proposed approaches I think are good to try:

1. Add maybe these optional CLI flags to `mlx_lm/lora.py`:
   - `--wired-limit-frac` (fraction of `max_recommended_working_set_size`)
   - `--memory-limit-frac` (same anchor)
   - `--cache-limit-frac` (same anchor)
Using a single anchor for all three keeps the interface predictable; anchoring `memory` and `cache` to `wired` would be more compact but harder to reason about.

2. Apply the limits inside `tuner/trainer.py:train()` before the loop starts, guarded by `mx.metal.is_available()` so non Metal builds are unaffected.

3. Document the flags in `LORA.md` with a short note on the memory/speed tradeoff (your "~80% speed at cache_limit=0" observation is useful here).

4. Add a brief tests that confirms the flags are respected like probably a monkeypatched `mx.set_*_limit` check rather than a real training run.

Two design questions before I open a PR, since the answers shape the interface:

(a) Should the defaults preserve current behavior (no limits set unless flags are explicitly passed) or apply safer values out-of-the-box. My instinct is the latter.

(b) Should this also cover `dora.py` and any other tuner entry points, or scope this PR to LoRA only and follow up separately for the other trainers?

Happy to adjust the scope or interface based on feedback.

**Comment by kru2710shna** [comment]
Opened https://github.com/ml-explore/mlx-lm/pull/1312 which exposes these as --wired-limit-ratio, --memory-limit-ratio, and --cache-limit-ratio on mlx_lm.lora (plus matching fields on TrainingArgs and yaml keys). Defaults preserve current behavior so it's a non-breaking change.

@simap if you get a chance, it'd be great to confirm the flags resolve your OOM scenario on your 48 GB setup.
